sa/sa.go

```diff
@@ -790,35 +790,63 @@ func (ssa *SQLStorageAuthority) FinalizeAuthorization2(ctx context.Context, req
 // RevokeCertificate stores revocation information about a certificate. It will only store this
 // information if the certificate is not already marked as revoked.
 func (ssa *SQLStorageAuthority) RevokeCertificate(ctx context.Context, req *sapb.RevokeCertificateRequest) (*emptypb.Empty, error) {
-	if req.Serial == "" || req.DateNS == 0 {
+	if req.Serial == "" || req.DateNS == 0 || req.IssuerID == 0 {
 		return nil, errIncompleteRequest
 	}
 
-	revokedDate := time.Unix(0, req.DateNS)
+	_, overallError := db.WithTransaction(ctx, ssa.dbMap, func(tx db.Executor) (interface{}, error) {
+		revokedDate := time.Unix(0, req.DateNS)
 
-	res, err := ssa.dbMap.ExecContext(ctx,
-		`UPDATE certificateStatus SET
+		res, err := tx.ExecContext(ctx,
+			`UPDATE certificateStatus SET
 				status = ?,
 				revokedReason = ?,
 				revokedDate = ?,
 				ocspLastUpdated = ?
 			WHERE serial = ? AND status != ?`,
-		string(core.OCSPStatusRevoked),
-		revocation.Reason(req.Reason),
-		revokedDate,
-		revokedDate,
-		req.Serial,
-		string(core.OCSPStatusRevoked),
-	)
-	if err != nil {
-		return nil, err
-	}
-	rows, err := res.RowsAffected()
-	if err != nil {
-		return nil, err
-	}
-	if rows == 0 {
-		return nil, berrors.AlreadyRevokedError("no certificate with serial %s and status other than %s", req.Serial, string(core.OCSPStatusRevoked))
+			string(core.OCSPStatusRevoked),
+			revocation.Reason(req.Reason),
+			revokedDate,
+			revokedDate,
+			req.Serial,
+			string(core.OCSPStatusRevoked),
+		)
+		if err != nil {
+			return nil, err
+		}
+		rows, err := res.RowsAffected()
+		if err != nil {
+			return nil, err
+		}
+		if rows == 0 {
+			return nil, berrors.AlreadyRevokedError("no certificate with serial %s and status other than %s", req.Serial, string(core.OCSPStatusRevoked))
+		}
+
+		if req.ShardIdx != 0 {
+			var rsm recordedSerialModel
+			err = tx.SelectOne(
+				ctx, &rsm, `SELECT expires FROM serials WHERE serial = ?`, req.Serial)
+			if err != nil {
+				return nil, fmt.Errorf("retrieving revoked certificate expiration: %w", err)
+			}
+
+			err = tx.Insert(ctx, &revokedCertModel{
+				IssuerID:      req.IssuerID,
+				Serial:        req.Serial,
+				NotAfter:      rsm.Expires,
```

Comment:
Looking through these fields, I can satisfy myself that each one will be non-zero, except for `NotAfter` / `rsm.Expires`, for which we're relying on the database not having that value as zero. Which I'm pretty confident is always the case... but perhaps we should 500 `rsm.Expires` is somehow zero (or 1970)?

Reply:
That's sufficient 👍🏻